core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -544,101 +550,204 @@ class EntityService(protected val ctx: RawlsRequestContext,
-          // update ENTITY from the contents of the temp table
-          _ = logger.info(s"Quicksilver migration: updating ENTITY from temp table ...")
-          _ <- dataAccess.compactEntityQuery.migrationUpdateFromTempTable(workspaceContext.workspaceIdAsUUID)
-          // populate the ENTITY_REFS table for this workspace
-          _ = logger.info(s"Quicksilver migration: populating ENTITY_REFS ...")
-          _ <- dataAccess.compactEntityQuery.migrationAddReferences(workspaceContext.workspaceIdAsUUID, shardId)
-          // delete legacy attributes from the ENTITY_ATTRIBUTE_xx_xx table
-          _ = logger.info(s"Quicksilver migration: deleting legacy attributes ...")
-          _ <- dataAccess.compactEntityQuery.migrationDeleteLegacyReferences(workspaceContext.workspaceIdAsUUID,
-                                                                             shardId
+        // finally, change the workspace to be quicksilver-enabled
+        _ <- traceFutureWithParent("setWorkspaceSettings", s) { _ =>
+          workspaceSettingService.setWorkspaceSettings(
+            workspaceName,
+            List(CompactDataTablesSetting(CompactDataTablesConfig(enabled = true)))
           )
-          // delete the all_attribute_values column for this workspace
-          _ = logger.info(s"Quicksilver migration: clearing all_attribute_values ...")
-          _ <- dataAccess.compactEntityQuery.migrationClearAllAttributesString(workspaceContext.workspaceIdAsUUID)
+        }
 
-          _ <- dataAccess.compactEntityQuery.migrationDeleteTempTable
-          _ = logger.info(s"Quicksilver migration: done!")
-        } yield ()
+        // return a count of entities updated
+      } yield userResult
+    }
+  }
+
+  private case class MigrationBoundary(startEntityId: Long, endEntityId: Long)
+
+  /**
+   * Calculate the boundaries for each batch of entities to migrate
+   * This method returns a sequence of MigrationBoundary objects, each containing the start and end entity ids
+   * for a batch. The batch sizes are determined by the batchSize parameter.
+   */
+  private def quicksilverCalculateBatches(workspaceId: UUID,
+                                          batchSize: Int,
+                                          dataAccess: DataAccess
+  ): ReadAction[Seq[MigrationBoundary]] = {
+
+    def findNextBatch(accum: Seq[MigrationBoundary], startingId: Long): ReadAction[Seq[MigrationBoundary]] =
+      dataAccess.compactEntityQuery.findMaxBatchId(batchSize, startingId, workspaceId).flatMap {
+        case None | Some(0L) => DBIO.successful(accum)
+        case Some(nextLimit) => findNextBatch(accum :+ MigrationBoundary(startingId, nextLimit), nextLimit)
       }
 
-      // finally, change the workspace to be quicksilver-enabled
-      _ <- workspaceSettingService.setWorkspaceSettings(
-        workspaceName,
-        List(CompactDataTablesSetting(CompactDataTablesConfig(enabled = true)))
-      )
+    findNextBatch(Seq.empty, -1)
+  }
```

Comment:
This recursive Scala function could be rewritten as a single SQL query which uses a [window function](https://dev.mysql.com/doc/refman/8.4/en/window-functions-usage.html). That query would look like:
```sql
WITH NumberedRows AS (
        SELECT
            id,
            ROW_NUMBER() OVER (ORDER BY id) AS row_num
        FROM ENTITY
        WHERE workspace_id = x'E9B3B3DCD8F24A88ACDB1D20C6A12032'
        and deleted = 0
      )
      SELECT
          FLOOR((row_num - 1) / 50000) AS win,
          MIN(id) AS min_id,
          MAX(id) AS max_id
      FROM NumberedRows
      GROUP BY win
      ORDER BY win;
```
I find the recursive Scala to be more readable; they are roughly equal in performance. Which would you rather see?

Reply:
I am not really a fan of db queries within a recursive loop and don't understand window functions. But I don't have a better suggestion that also minimizes data transfer. I.e. a batched streaming approach sounds good but we don't want to stream the data through the app.

So I guess stick with what you have.